feat: `docker-build-push`

marek-saji opened the pull request:
# Why?

Closes https://verkstedt.atlassian.net/browse/VIP-48

# What?

Workflow that is triggered when new version tag or anything to main is pushed. It will build a docker image and depending on vars set in the repo (or org), will push it to different registries (GitHub Container Registry, Google Artifact Registry, Amazon ECR).

# Review notes

> [!IMPORTANT]
> Remember this is a public repo. Do not discuss internal projects.

- Pushing to Amazon has _NOT_ been tested.
- Vars can be seen at https://github.com/verkstedt/actions--test/settings/variables/actions

## Testing

There’s a private test repo set up at https://github.com/verkstedt/actions--test it sends notifications to `#infrastructure-create-release-pr-action` slack channel.

# TODO before this is merged

- [x] Replace all `@feat/docker-build-push` with `@v1`

# TODO after this is merged

- [x] Add workflow template to `verkstedt/.github` repo. Will look something like https://github.com/verkstedt/actions--test/blob/main/.github/workflows/docker-build-push.yaml
- [x] Stop using `reusable-tag-docker-release-images` and `reusable-release` workflows and remove them
- [ ] https://verkstedt.atlassian.net/browse/RDC999-406

[VIP-48]: https://verkstedt.atlassian.net/browse/VIP-48?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

nuc committed:
docker-build-push

nuc committed:
docker-build-push: Pass build_args to the build

nuc committed:
docker-build-push: Pass build_args as secret

nuc committed:
docker-build-push: Define secrets in the workflow

marek-saji committed:
docker-build-push: Build without any tags

Refs https://github.com/verkstedt/actions/pull/20#discussion_r2222058417

marek-saji committed:
docker-build-push: Language

Refs https://github.com/verkstedt/actions/pull/20#discussion_r2222058427

marek-saji committed:
docker-build-push: Define version_change output

marek-saji committed:
docker-build-push: Validate required inputs

marek-saji committed:
create-release-pr: Pretty white space in CHANGELOG

marek-saji committed:
feat(create-release): Use semantic-release

marek-saji committed:
docker-build-push: Push to all registries in one go

marek-saji committed:
chore: Kebab case for inputs

marek-saji committed:
docker-build-push

marek-saji committed:
create-release-pr

marek-saji committed:
create-release: new vision

marek-saji committed:
docker-build-push: new vision

marek-saji committed:
docker-build-push: Update actions

marek-saji committed:
create-release: Remove working-dir input, until it’s needed

marek-saji committed:
create-release: dry-run

marek-saji committed:
create-release: Triggering GH user in git commit

marek-saji committed:
create-release: Use ::error::

marek-saji committed:
fix(create-release): Piping release notes to step summary

marek-saji committed:
create-release: Notify about success

marek-saji committed:
create-release: No notify on dry-run

marek-saji committed:
create-release: Use GitHub app token

marek-saji committed:
Remove TODOs

marek-saji committed:
doc(create-release)

marek-saji committed:
docker-build-push: kebab-case for all inputs

marek-saji committed:
docker-build-push: Notify about success

marek-saji committed:
docker-build-push

marek-saji committed:
docker-build-push: Build and push in one step

marek-saji committed:
DEBUG(create-release): Use setup@feat/create-release

marek-saji committed:
docker-build-push: More pleasant version tag order

marek-saji committed:
docker-build-push: Checkout before auth

marek-saji committed:
create-release: Note about $s

marek-saji committed:
create-release: Skip previous_tag_name, if there’s none

marek-saji committed:
create-release: Custom success PR comment

marek-saji committed:
fixup! create-release: Skip previous_tag_name, if there’s none

marek-saji committed:
fixup! create-release: Note about $s

marek-saji committed:
fixup! create-release: Custom success PR comment

marek-saji committed:
create-release: unindent shell commands

marek-saji committed:
fixup! create-release: Custom success PR comment

marek-saji committed:
fixup! create-release: Use GitHub app token

marek-saji committed:
docker-build-push: Change most inputs to vars

marek-saji committed:
create-release: Improve setup instructions

marek-saji committed:
docker-build-push: No _ in domain name

marek-saji committed:
docker-build-push: edge tag always on main branch

marek-saji committed:
docker-build-push: Markdown headings

marek-saji committed:
docker-build-push: Show platforms in summary

marek-saji committed:
docker-build-push: Example for platforms input

marek-saji committed:
docker-build-push: Notify about succes only when tagging with latest

marek-saji committed:
docker-build-push: tweak

marek-saji committed:
create-release: docs

marek-saji committed:
fixup! docker-build-push: Change most inputs to vars

Resolves https://github.com/verkstedt/actions/pull/26#discussion_r2428888065
Resolves https://github.com/verkstedt/actions/pull/26#discussion_r2428888077

marek-saji committed:
fixup! docker-build-push

Resolves https://github.com/verkstedt/actions/pull/26#discussion_r2428888083

smelchior commented on .github/workflows/docker-build-push.yaml:
```diff
@@ -0,0 +1,369 @@
+        description: 'Build arguments for Docker build (multiline string)'
+      SLACK_BOT_TOKEN:
+        description: 'Slack bot token, to notify a Slack channel of successful run'
+        required: true
```
Why do we have to enforce sending something to slack?

marek-saji replied:
Hm. Perhaps I was thinking about it from perspective of a project I was working on where we push to registry and then notify the client that new version is available there.

How does it look in other projects? Do you do releases often enough that these messages would be annoying?

If it’s not that bad, I’d keep it for the sake of consistency between projects, but I’m up for discussing it. On slack maybe, so we can get into details.

smelchior replied:
Well for example in the ti-hub projects a new release is not notified to the client as we take care of the deployment in another repository ourselves. So for this i would guess it is a bit annoying.

marek-saji replied:
I’d like to keep it on by default, for consistency. What do you think about introducing something like a `DOCKER_SLACK_NOFICATION=false` var?

smelchior commented on .github/workflows/docker-build-push.yaml:
```diff
@@ -0,0 +1,369 @@
+        type: boolean
+        default: false
+      platforms:
+        description: 'Platforms to build for, e.g. linux/amd64,linux/arm64'
```
could we add the linux/amd64 as a default here and drop the DOCKER_PLATFORMS env var?

marek-saji replied:
My thinking was that some projects might want to build by default for other platforms. You don’t think that’s necessary? I guess we can always bring it back, if it turns out to be useful in the future.

smelchior replied:
Well if they want it they can set the config for the build mode i think? I do not currently see a usecase of this flexibility, but we can keep it.

marek-saji replied:
I’ll drop it. Better to start with a simpler version. We can always re–introduce it, when there’s a need.

smelchior commented on .github/workflows/docker-build-push.yaml:
```diff
@@ -0,0 +1,369 @@
+            exit 64 # EX_USAGE
+          fi
+
+      - name: 'Determine platforms and tags to use'
```
I think we should add an option to somehow select what tags to use
The problem here is that some registries might be configured as immutable -> therefore latest/edge might not be used as tag names. 
Maybe we can have a registry_immutable argument to decide this + an optional tags argument to just pass the full tags list and override this logic?

marek-saji replied:
Do you mean that some regs in e.g. google registry might not allow to update tags, or that whole platform (e.g. amazon) wouldn’t allow it?

smelchior replied:
That depends on the configuration of the repository. That can either be GCP or AWS or .. 
For some it might be enabled as a best practice for security or required by some regulation.

marek-saji replied:
Do you think we need to make it configurable per registry (i.e. separately for github, google and amazon) or is it fine to have one setting per repo?

Per repo makes more sense to me, since it gives to the same set of tags for each provider.

marek-saji replied:
Though, following logic from https://github.com/verkstedt/actions/pull/26#discussion_r2510160573 — do we already have a project that would require this? Or maybe we can add it  when the need arise?

smelchior replied:
For the projects in the ti-hub context it is the case which are pushed to the ECR of Arvato

marek-saji committed:
fixup! create-release: Triggering GH user in git commit

Resolves https://github.com/verkstedt/actions/pull/25#discussion_r2452355090

marek-saji committed:
create-release: Move config to separate file

Resolves https://github.com/verkstedt/actions/pull/25#discussion_r2452370872

marek-saji committed:
Merge remote-tracking branch 'origin/v1' into feat/create-release

marek-saji committed:
create-release: New version and notes in notification

marek-saji committed:
fixup! create-release: New version and notes in notification

Resolves https://github.com/verkstedt/actions/pull/25#discussion_r2455492486

marek-saji committed:
fixup! create-release: New version and notes in notification

Resolves https://github.com/verkstedt/actions/pull/25#discussion_r2455492468

marek-saji committed:
create-release: Quote shell env vars

Resolves https://github.com/verkstedt/actions/pull/25#discussion_r2455561392, https://github.com/verkstedt/actions/pull/25#discussion_r2455561411 and https://github.com/verkstedt/actions/pull/25#discussion_r2455561422

marek-saji committed:
docker-build-push: Drop DOCKER_PLATFORMS var

Resolves https://github.com/verkstedt/actions/pull/26#discussion_r2452390416

marek-saji committed:
docker-build-publish: Option to not set mutable tags

Resolves https://github.com/verkstedt/actions/pull/26#discussion_r2452399652

marek-saji commented:
@coderabbitai pause

marek-saji committed:
docker-build-push: Option to skip Slack notifiactions

Resolves https://github.com/verkstedt/actions/pull/26#discussion_r2452380811

marek-saji committed:
setup: Expand token–related descriptions

Resolves https://github.com/verkstedt/actions/pull/25#discussion_r2498287466
Resolves https://github.com/verkstedt/actions/pull/25#discussion_r2498298811

marek-saji committed:
setup: Multi–line descriptions

marek-saji committed:
docker-build-push: WIP inputs, not vars

marek-saji committed:
Typo

marek-saji committed:
create-release: inputs, not vars

marek-saji committed:
docker-build-push: Inputs, not vars

marek-saji committed:
docker-build-push: Extract provider setups

marek-saji committed:
create-release: Drop ENV_VARS

marek-saji committed:
fixup! docker-build-push: Extract provider setups

smelchior commented on .github/workflows/docker-build-push.yaml:
```diff
@@ -0,0 +1,340 @@
+        uses: verkstedt/actions/docker-setup-registry/github@feat/docker-build-push
+        with:
+          image-name: ${{ inputs.github-image-name }}
+          github-token: ${{ secrets.GITHUB_TOKEN }}
```
Q: should we allow to set this if you need to update a ghcr registry from another entity where the local token will not work?

marek-saji replied:
I don’t _think_ that’s necessary, but let’s add it to avoid breaking changes in the future.

Added in https://github.com/verkstedt/actions/commit/39eb984edc8c9de95cb7b069f5a75b0e745fbce2

smelchior replied:
Can we make secrets.GITHUB_TOKEN the default? So if nothing is passed, that is what is used?

marek-saji replied:
Hm, I’ll test it out. If we can do it, then we could drop the input secret altogether, because introducing it later wouldn’t be a breaking change.

marek-saji replied:
Yeah. Let’s leave it out and keep using `secrets.GITHUB_TOKEN`. If we ever need to use a different token, we can add a secret input then and use it as `github-token: ${{ secrets.GHCR_GITHUB_TOKEN || secrets.GITHUB_TOKEN }}`.

smelchior commented on README.md:
```diff
@@ -122,6 +173,15 @@ Template:
+
+Template:
+<https://github.com/verkstedt/.github/tree/main/workflow-templates/docker-build-push.yaml>
+
```
template is not present (yet)

marek-saji replied:
See “TODO after this is merged” in the PR body :slightly_smiling_face:

marek-saji committed:
fixup! fixup! docker-build-push: Extract provider setups

marek-saji committed:
fixup! create-release: New version and notes in notification

marek-saji committed:
feat(notify-status): Include text–only value

For OS notifications and such.

marek-saji committed:
feat(create-release): Graceful when slackifying fails

marek-saji committed:
fixup! feat(notify-status): Include text–only value

marek-saji committed:
fixup! fixup! feat(notify-status): Include text–only value

stepahn commented on .github/workflows/docker-build-push.yaml:
```diff
@@ -0,0 +1,347 @@
+          } >> "$GITHUB_STEP_SUMMARY"
+
+          # `git-<SHORT_SHA>`
+          TAGS="git-$( git rev-parse --short HEAD )"
```
how about including an additional tag based on the commit timestamp which yields an automatic monotonous increasing "version" to use in situations where you have no semantic version?

```
 CT=$(git show -s --format=%ct HEAD)
 echo "COMMIT_TIMESTAMP=$(date -u -d "@$CT" +%Y%m%d%H%M%S)" >> $GITHUB_OUTPUT
```

marek-saji replied:
Will do. Btw. git has `--date=format:…`, so we can do it in one go.

marek-saji replied:
@stepahn I’d like to go with `date-<DATE>`. Is that fine with you?

marek-saji replied:
And a fixup in dc795a267f9c98e4e166fda735f30a42f88a85db

stepahn replied:
sure, that's fine

stepahn commented on .github/workflows/docker-build-push.yaml:
```diff
@@ -0,0 +1,347 @@
+          platforms: ${{ steps.config.outputs.platforms }}
+          provenance: true
+          tags: ${{ steps.config.outputs.tags }}
+          build-args: ${{ secrets.BUILD_ARGS }}
```
might be useful to pass `secrets` as well

```
    secrets: |
      GITHUB_TOKEN=${{ some.other.job.provides.a.secret.token }}
```

if you need to fetch private resources to build the docker image

marek-saji replied:
What do you think about changing `BUILD_ARGS` secret into an input and introducing `secrets.BUILD_SECRETS` instead?

stepahn replied:
That's fine, I'll give it a try

stepahn commented:
looks good to me, I'll give the registry actions a try and see if I can adopt those in TIH though I'm not sure if I can use the whole docker-build-push action though

marek-saji commented:
> looks good to me, I'll give the registry actions a try and see if I can adopt those in TIH though I'm not sure if I can use the whole docker-build-push action though

Maybe at least the `docker-setup-registry/*`.

marek-saji committed:
feat(docker-build-push): `date-<DATE>` docker tag

Resolves https://github.com/verkstedt/actions/pull/26#discussion_r2536405792

marek-saji committed:
feat(docker-build-push): Build args and secrets

Resolves https://github.com/verkstedt/actions/pull/26#discussion_r2536411138

marek-saji committed:
feat(docker-build-push): Fall back GHCR_GITHUB_TOKEN to GITHUB_TOKEN

marek-saji committed:
fixup! feat(docker-build-push): `date-<DATE>` docker tag

marek-saji committed:
feat(docker-build-push): Drop GHCR_GITHUB_TOKEN

Using build–in secrets.GITHUB_TOKEN. If we’ll ever need an option to use
different token, we can introduce that in a non–breaking way.

Resolves https://github.com/verkstedt/actions/pull/26#discussion_r2536909176

stepahn commented:
> > looks good to me, I'll give the registry actions a try and see if I can adopt those in TIH though I'm not sure if I can use the whole docker-build-push action though
> 
> Maybe at least the `docker-setup-registry/*`.

yeah, the registry actions I'm gonna try … and with the recent additions I can try and see if I could use the whole aciton

marek-saji committed:
feat(create-release): Update package.json if exists only

marek-saji committed:
fix(create-release): Drop type:module import assertion

We already run whole node with --input-type=module

marek-saji committed:
fixup! feat(create-release): Update package.json if exists only

marek-saji committed:
feat(create-release): Check other workflows before running

marek-saji committed:
Merge remote-tracking branch 'origin/v1' into feat/create-release

marek-saji committed:
Merge remote-tracking branch 'origin/v1' into feat/docker-build-push

marek-saji committed:
fixup! feat(create-release): Check other workflows before running

marek-saji committed:
fix(create-release): Brute force when installing semantic-release plugins

marek-saji committed:
feat(create-release): Split Release step

marek-saji committed:
fixup! docker-build-push: Inputs, not vars

marek-saji committed:
docker-build-push: Drop vMAJOR and vMAJOR.MINOR tags

marek-saji committed:
Merge remote-tracking branch 'origin/v1' into feat/create-release

marek-saji committed:
chore(create-release): Switch versions from feat branch to v1

marek-saji committed:
chore: prettier -w .

marek-saji committed:
Merge remote-tracking branch 'origin/feat/create-release' into feat/docker-build-push

marek-saji committed:
chore(create-release): Switch versions from feat branch to v1
